### PR TITLE
Use std::initializer_list for braced-list construction

### PR DESCRIPTION
**Pull request by pazner**
In the previous implementation, `Array<T>({1})` would select the `Array<T>(int)` constructor rather than the braced-list constructor, for example:

```
warning: braces around scalar initializer [-Wbraced-scalar-init]
      |    Array<int> array({1});
```

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4564](https://github.com/mfem/mfem/pull/4564) | @pazner | @v-dobrev | @paul-hilscher + @najlkin | 11/1/24 | 11/21/24 | 11/25/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

File: general/array.hpp [hunk 1]
```diff
@@ -92,8 +93,8 @@ class Array
    inline Array(const Array<CT> &src);
 
    /// Deep copy from a braced init-list of convertible type
-   template <typename CT, int N>
-   explicit inline Array(const CT (&values)[N]);
+   template <typename CT>
+   explicit inline Array(std::initializer_list<CT> values);
```
Comment (v-dobrev) [inline general/array.hpp:97]: I think we should keep the original implementation -- it allows construction which is not allowed by the new version.

Comment (pazner) [reply general/array.hpp]: You mean initializing the Array with a C-style array with static size? I re-added the previous one (in addition to the new one) in c7bb0201c0745a17c4a37874315a91a942e49a6f.

File: linalg/vector.hpp [hunk 2]
```diff
@@ -120,9 +121,10 @@ class Vector
       : data(size_, h_mt, d_mt), size(size_) { }
 
    /// Create a vector using a braced initializer list
-   template <int N, typename T = real_t>
-   explicit Vector(const T (&values)[N]) : Vector(N)
-   { std::copy(values, values + N, GetData()); }
+   template <typename CT, typename std::enable_if<
+                std::is_convertible<CT,real_t>::value,bool>::type = true>
+   explicit Vector(std::initializer_list<CT> values) : Vector(values.size())
+   { std::copy(values.begin(), values.end(), begin()); }
```
[the remark this answers is not in the document]
Comment (pazner) [reply linalg/vector.hpp:132]: Done in b4efd401b8c9e11135ed04e231c895d0c9dbe59b.

File: linalg/vector.hpp [hunk 3]
```diff
@@ -121,7 +121,7 @@ class Vector
       : data(size_, h_mt, d_mt), size(size_) { }
 
    /// Create a vector from a statically sized C-style array of convertible type
-   template <int N, typename CT>
+   template <typename CT, int N>
```
Comment (v-dobrev) [inline linalg/vector.hpp:124]: This is not exactly backward-compatible -- e.g. when using explicit template parameters.

Comment (pazner) [reply linalg/vector.hpp:124]: True, but I suspect none of our users have explicitly specified them. We can revert the order if you want though. I changed it for consistency with Array, but it's not so important.

Comment (v-dobrev) [reply linalg/vector.hpp:124]: Yes, it is not very likely that someone used explicit template parameters. I'll leave it up to you to decide if you want to revert -- I'm fine either way.

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: tests/unit/general/test_array.cpp [hunk 4]
```diff
@@ -14,15 +14,17 @@
 
 using namespace mfem;
 
-TEST_CASE("Array init-list construction", "[Array]")
+TEST_CASE("Array init-list and C-style array constructors", "[Array]")
 {
    int ContigData[6] = {6, 5, 4, 3, 2, 1};
    Array<int> a(ContigData, 6);
-   Array<int> b({6.0, 5.0, 4.0, 3.0, 2.0, 1.0});
```
Comment (najlkin) [inline tests/unit/general/test_array.cpp:21]: This is no longer possible? If implicit convertibility is kept, it should be, so I would intentionally keep/add this version of the test. The same thing for `Vector` below in the opposite way.

File: linalg/vector.hpp [hunk 5]
```diff
@@ -119,10 +120,16 @@ class Vector
...
+   explicit Vector(const CT (&values)[N]) : Vector(N)
+   { std::copy(values, values + N, begin()); }
+
    /// Create a vector using a braced initializer list
-   template <int N, typename T = real_t>
-   explicit Vector(const T (&values)[N]) : Vector(N)
-   { std::copy(values, values + N, GetData()); }
+   template <typename CT, typename std::enable_if<
+                std::is_convertible<CT,real_t>::value,bool>::type = true>
```
Comment (najlkin) [inline linalg/vector.hpp:130]: This check would be nice also for `Array` maybe 😉 .

File: tests/unit/general/test_array.cpp (hunk 4 above)
Comment (pazner) [reply tests/unit/general/test_array.cpp:21]: This is still possible. I assumed that the `.0` was just a typo, but we can re-add if it was intentional.

Comment (najlkin) [reply tests/unit/general/test_array.cpp:21]: Maybe was 😆  , but would be nice as an extra test for the convertibility.

Comment (pazner) [reply tests/unit/general/test_array.cpp:21]: Re-added in c9cf2ec345f9514506af5da5f8002c5ab14adb21.

File: tests/unit/linalg/test_vector.cpp [hunk 6]
```diff
@@ -15,15 +15,17 @@
 
 using namespace mfem;
 
-TEST_CASE("Vector init-list construction", "[Vector]")
+TEST_CASE("Vector init-list and C-style array constructors", "[Vector]")
 {
    real_t ContigData[6] = {6.0, 5.0, 4.0, 3.0, 2.0, 1.0};
    Vector a(ContigData, 6);
    Vector b({6.0, 5.0, 4.0, 3.0, 2.0, 1.0});
```
Comment (najlkin) [inline tests/unit/linalg/test_vector.cpp:22]: ☝️  Here without `.0`.

File: general/array.hpp [hunk 7]
```diff
@@ -23,6 +23,7 @@
 #include <cstring>
 #include <algorithm>
 #include <type_traits>
+#include <initializer_list>
```
Comment (paul-hilscher) [inline general/array.hpp:26]: minor: `#include <type_traits>` due to `std::enable_if` and `std::initializer_list` ?

File: general/array.hpp [hunk 8]
```diff
@@ -91,10 +92,15 @@ class Array
...
    inline Array(const Array<CT> &src);
 
-   /// Deep copy from a braced init-list of convertible type
+   /// Construct an Array from a C-style array of static length
    template <typename CT, int N>
    explicit inline Array(const CT (&values)[N]);
 
+   /// Construct an Array from a braced initializer list of convertible type
+   template <typename CT, typename std::enable_if<
+                std::is_convertible<CT,T>::value,bool>::type = true>
```
Comment (paul-hilscher) [inline general/array.hpp:101]: minor: mfem-4.8 should support C++17 (based on Tzanio's _state of mfem_ workshop slides) ; if we can already use it maybe ` std::enable_if_t<std::is_convertible_v<CT, T>>` might be simpler.

**Review by paul-hilscher — APPROVED** [review]
lgtm. I have added two minor points/questions.

File: tests/unit/linalg/test_vector.cpp (hunk 6 above)
Comment (pazner) [reply tests/unit/linalg/test_vector.cpp:24]: Done in c232029d84270e0ec0fee02aea5efb7e3719ff7d.

File: general/array.hpp (hunk 7 above)
Comment (pazner) [reply general/array.hpp:26]: It's already included in this header, but I added the include to `vector.hpp` in c232029d84270e0ec0fee02aea5efb7e3719ff7d.

File: general/array.hpp (hunk 8 above)
Comment (pazner) [reply general/array.hpp:101]: @tzanio, thoughts?

**Review by najlkin — APPROVED** [review]
LGTM, thanks for cleaning this up 😉 .

File: general/array.hpp (hunk 8 above)
Comment (v-dobrev) [reply general/array.hpp:101]: Since using C++11 is possible and not much more complicated, I'd say, let's stick with C++11 for this.

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Pull request merged**